### PR TITLE
Health simplification

### PR DESCRIPTION
**Pull request by Bod9001**
chemicals now use tags to determine body parts, 
removed blood from everybody part being present,
centralised the calculation is in Circulatory  system

### Purpose
to  Improve performance of health system

### Changelog:

 CL: [Improvement] Improved performance of health system

### CODE REVIEW

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs [hunk 1]
```diff
@@ -272,41 +278,41 @@ public void OxyHeal(ReagentMix reagentMix, float amount)
...
-
-				BloodContainer.CurrentReagentMix.Add(Nutriment, availableNutriment);
-			}
-			else
-			{
-				HungerModifier.Multiplier = 0.5f;
-				HungerState = HungerState.Starving;
-				// Is Starving
-			}
+			// float availableNutriment = BloodContainer.CurrentReagentMix.Subtract(Nutriment, Single.MaxValue);
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs]: Don't comment out things, just delete them

**Comment by Bod9001** [comment]
probably need to do a bit more testing

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 2]
```diff
@@ -1443,6 +1416,41 @@ private void FastRegen()
...
+				InstantiateAndSetUp(RaceBodyparts.Base.Head);
+				InstantiateAndSetUp(RaceBodyparts.Base.Torso);
+				InstantiateAndSetUp(RaceBodyparts.Base.ArmLeft);
+				InstantiateAndSetUp(RaceBodyparts.Base.ArmRight);
+				InstantiateAndSetUp(RaceBodyparts.Base.LegLeft);
+				InstantiateAndSetUp(RaceBodyparts.Base.LegRight);
+			}
+		}
+
+		public void InitialiseFromRaceData(PlayerHealthData RaceBodyparts)
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:1433]: How does this handle races that may not have metabolism, blood or toxins?

File: UnityProject/Assets/Scripts/Items/ItemAttributesV2.cs [hunk 3]
```diff
@@ -247,7 +247,14 @@ public bool HasAnyTrait(IEnumerable<ItemTrait> expectedTraits)
 		/// <returns></returns>
 		public bool HasAllTraits(IEnumerable<ItemTrait> expectedTraits)
 		{
-			return traits.All(expectedTraits.Contains);
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Items/ItemAttributesV2.cs:250]: why this change?

File: UnityProject/Assets/Scripts/Items/Implants/Organs/Heart.cs [hunk 4]
```diff
@@ -55,16 +52,27 @@ public override void ImplantPeriodicUpdate()
 			DoHeartBeat();
 		}
 
+		public override void RemovedFromBody(LivingHealthMasterBase livingHealth)
+		{
+			livingHealth.CirculatorySystem.Hearts.Remove(this);
+		}
+
+		public override void AddedToBody(LivingHealthMasterBase livingHealth)
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Items/Implants/Organs/Heart.cs:60]: this function's name (as well as `RemovedFromBody()` make it sound like it's a bool rather than a void, `AddToBody()` would be a clearer name though this is just a nitpick.

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 5]
```diff
@@ -402,5 +402,10 @@ private void UpdateIcons()
 		{
 			UIManager.PlayerHealthUI.SetBodyTypeOverlay(this);
 		}
+
+		public float GetDamage(DamageType damageType)
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:406]: why is this float instead of an int

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs [hunk 6]
```diff
@@ -158,11 +153,13 @@ public void BloodInitialise()
...
 		{
+			return;
+
 			if (isBloodCirculated == false) return;
-			ConsumeReagents();
+			ConsumeReagents(); //d
 			if (CanGetHungry)
 			{
-				ConsumeNutriments();
+				ConsumeNutriments(); //d
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs]: //d ?

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs [hunk 7]
```diff
@@ -84,13 +84,8 @@ public partial class BodyPart
...
 
-		public enum RequiredBlood
-		{
-			Large = 100,
-			Medium = 20,
-			Low = 10,
-			Minuscule = 1
-		}
+
+		public float currentBloodSaturation = 0;
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs:76]: private variable?

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BloodTypes/BloodType.cs [hunk 8]
```diff
@@ -55,5 +65,30 @@ public float GetSpareGasCapacity(ReagentMix reagentMix, Reagent reagent = null)
 			}
 			return GetGasCapacity(reagentMix, reagent) - reagentMix[reagent];
 		}
+
+
+		public float CalculatePercentageBloodPresent(ReagentMix reagentMix)
+		{
+			var Percentage = (reagentMix[this] / reagentMix.Total);
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BloodTypes/BloodType.cs:72]: what if `this` is null in the `reagentMix` list? or is it impossible to be able to do that?

File: UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs [hunk 9]
```diff
@@ -42,56 +40,116 @@ public struct TypeAndStrength
...
+						foreach (var Effect in Effects)
+						{
+							bodyPart.TakeDamage(null,
+								Effect.EffectPerOne * TotalChemicalsProcessedByBodyPart * -OverdoseDamageMultiplier,
+								Effect.AttackType,
+								Effect.DamageEffect, DamageSubOrgans: false);
+						}
+					}
+					else
+					{
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs:126]: can be inverted

File: UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/MetabolismReaction.cs [hunk 10]
```diff
@@ -24,44 +31,57 @@ public override bool Apply(MonoBehaviour sender, ReagentMix reagentMix)
 			return false;
 		}
 
-		var bodyPart = sender.GetComponent<BodyPart>();
+		var bodyPart = sender.GetComponent<CirculatorySystemBase>();
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/MetabolismReaction.cs]: wrong variable name?

File: UnityProject/Assets/Scripts/Items/ItemAttributesV2.cs (hunk 3 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Items/ItemAttributesV2.cs:250]: it required having the exact same list of ItemTrait for some reason

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 5 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:406]: all damages float

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs (hunk 7 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs:76]: is needed for organ functionality

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BloodTypes/BloodType.cs (hunk 8 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BloodTypes/BloodType.cs:72]: it's impossible

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs (hunk 7 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BodyPartBlood.cs:76]: then captalize the first C or make it private and make a public getter

File: UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BloodTypes/BloodType.cs (hunk 8 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/CirculatorySystem/BloodTypes/BloodType.cs:72]: alright cool

File: UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs (hunk 9 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs:126]: oh why so?

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:1433]: doesn't currently but it is easy enough to add

File: UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs (hunk 9 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs:126]: clean code

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:1433]: would be good to have

File: UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs (hunk 9 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/MedicalChemistry/BodyHealthEffect.cs:126]: why so?
